### PR TITLE
mp_create_game: Show preview if map_file is set

### PR DESCRIPTION
**Pull request by jostephd**
Fixes the second bullet of https://github.com/wesnoth/wesnoth/issues/4397#issuecomment-536357310.

Could someone double check the change to the generator_assigned condition? Is it about the case that a scenario file defines both map_file and a map generator? I would have expected that to be an error, actually.

### CODE REVIEW

**Comment by gfgtdf** [comment]
> Could someone double check the change to the generator_assigned condition?

> Is it about the case that a scenario file defines both map_file and a map generator? 

yes.

The result of the map generator is  stored in the map_data attribute , so that code checks whether the map was already generated and does so if it didnt happen yet.

**Comment by gfgtdf** [comment]
from looking  at the code im suprised this works, since all codes you are changing are inside `type == ng::level::TYPE::RANDOM_MAP` blocks, and i'd expect that the normal mp maps are not 'random maps' in fact random maps are the only type of maps where map_file does not make sense.

**Comment by jostephd** [comment]
@gfgtdf That branch of the `switch` statements handles all of the following:

```
		case ng::level::TYPE::SCENARIO:
		case ng::level::TYPE::USER_MAP:
		case ng::level::TYPE::USER_SCENARIO:
		case ng::level::TYPE::RANDOM_MAP: {
```

**Comment by gfgtdf** [comment]
> @gfgtdf That branch of the switch statement handles all of the following:

right, forogt how C++ switch works for a second.

but don't we use FALLTHROUGH for intentional fallthough switches and don't he compiler give a wanring for missing FALLTHROUGH?

**Comment by jostephd** [comment]
> but don't we use FALLTHROUGH for intentional fallthough switches and don't he compiler give a wanring for missing FALLTHROUGH?

Compilers probably suppress that warning when a label is followed by another label with no statements in between. Anyway, falling through happens whenever there's no `break;` statement; comments don't affect control flow.

**Comment by jostephd** [comment]
By the way, I wanted to put somewhere an accessor function, foo::map_data() that just gets the map_data from either the `map_data` attribute or the `map_file` attribute, whichever exists, and use that accessor function in all three places that currently access `"map_file"`...but I couldn't figure out where to put the new function. The three places that handle map_file don't seem to even have a common #include.

**Comment by soliton-** [comment]
Factoring the code out to a common function sounds good. Not sure where to put it either though.

**Comment by jostephd** [comment]
I could just create a non-member function somewhere and have the three places use it. Even if that's not the best possible place to declare the function in, we can move it to a more suitable home when we come up with one.

CC @Vultraz

**Comment by gfgtdf** [comment]
well at least one of three function that use map_file don't 'access' it they, read map_file and store it into map_data. So not sure how muhc there is in common actually.

**Comment by jostephd** [comment]
> well at least one of three function that use map_file don't 'access' it they, read map_file and store it into map_data. So not sure how muhc there is in common actually.

I thought of a function that takes a config object and returns a string. That string will be `c.has_attribute("map_file") ? filesystem::read_map(c["map_file"]) : c["map_data"]`. All three locations can use that, I think.

**Comment by gfgtdf** [comment]
hmm no the saved_game code does more like `if c.has_attribute("map_file") then c["map_data"] = read_map(c["map_file"]) end return c["map_data"];` and in fact i think that'd also make more sense for the mp_create code so that the map_file is not read twice.

**Comment by jostephd** [comment]
> and in fact i think that'd also make more sense for the mp_create code so that the map_file is not read twice.

Done

File: data/multiplayer/scenarios/2p_Aethermaw.cfg [hunk 1]
```diff
@@ -4,7 +4,7 @@
     name= _ "2p — Aethermaw"
     # wmllint: local spellings Sulla Aethermaw Paterson
     description= _ "Long ago, the Great Mage Sulla was imprisoned in the Aethermaw, a nexus of mystical energy whose chaotic nature was able to prevent her escape. Over the centuries, however, Sulla gradually attuned her powers to the maelstrom of disorder that is the Aethermaw, and has now begun to project its influence onto the material plane, drawing in entire regions of land from hundreds of different worlds, realities and time-periods. She experiments with these disparate pieces of the cosmos, manipulating them, merging them and sending them back and forth between the Aethermaw and their place of origin. Perhaps, as her mastery over the Aethermaw grows, Sulla will one day break free of its bonds. Until that time comes, she will continue to amuse herself by arranging battles between the mortal beings unlucky enough to be drawn into its depths. Designed by Doc Paterson."
-    map_data="{multiplayer/maps/2p_Aethermaw.map}"
+    map_file=multiplayer/maps/2p_Aethermaw.map
```
Comment (jostephd) [inline data/multiplayer/scenarios/2p_Aethermaw.cfg]: `error validation: Invalid key 'map_file=' in tag [multiplayer]`

Comment (jostephd) [reply data/multiplayer/scenarios/2p_Aethermaw.cfg]: The rebase should fix this.

**Pull request merged**